# Notebook 01: Data Acquisition and Understanding

## Objective
The objective of this notebook is to import the research dataset, understand its structure, identify the available variables, and verify that the data has been imported correctly before any preprocessing is performed.

In [1]:
# import libraries
import pandas as pd
import numpy as np

from pathlib import Path

| Dataset                      | Source                     | Expected Role            |
| ---------------------------- | -------------------------- | ------------------------ |
| HistoricalRateDetail         | South African Reserve Bank | Exchange rate (USD/ZAR)  |
| CPI_Average Prices_All urban | Statistics South Africa    | National food prices     |
| CPI_Average Prices_Provinces | Statistics South Africa    | Provincial food prices   |
| CPI (COICOP 2018 - 8 digit)  | Statistics South Africa    | Detailed food categories |


In [2]:
# load the dataset
raw_data_directory = Path("../data/raw")

exchange = pd.read_csv(
    raw_data_directory / "HistoricalRateDetail.csv"
)

prices_urban = pd.read_excel(
    raw_data_directory
    / "CPI_Average Prices_All urban(202604).xlsx"
)

prices_provinces = pd.read_excel(
    raw_data_directory
    / "CPI_Average prices_Provinces(202604).xlsx"
)

cpi_coicop = pd.read_excel(
    raw_data_directory
    / "EXCEL - CPI (COICOP 2018 - 8digit) (202604).xlsx"
)

## Inspect HistoricalRateDetail

In [3]:
#view the data
exchange.head(20)

,Indicator,Description
0,Rand per US Dollar,Weighted average of the banks' daily rates at ...
1,Date,Value
2,2026-05-18,16.6453
3,2026-05-15,16.6666
4,2026-05-14,16.4063
5,2026-05-13,16.4604
6,2026-05-12,16.5431
7,2026-05-11,16.4093
8,2026-05-08,16.4266
9,2026-05-07,16.2942


In [4]:
exchange.info()
exchange.describe()

<class 'pandas.DataFrame'>
RangeIndex: 2843 entries, 0 to 2842
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   Indicator    2843 non-null   str  
 1   Description  2843 non-null   str  
dtypes: str(2)
memory usage: 44.6 KB


,Indicator,Description
count,2843,2843
unique,2843,2777
top,Rand per US Dollar,14.4599
freq,1,3


In [5]:
exchange.tail()

,Indicator,Description
2838,2015-01-08,11.6426
2839,2015-01-07,11.7248
2840,2015-01-06,11.7264
2841,2015-01-05,11.7071
2842,2015-01-02,11.6219


### Observations of the HistoricalRateDetail dataset

The HistoricalRateDetail dataset was imported with metadata included in the first two rows. The first row contains a description of the exchange-rate series, while the second row contains the actual variable names ("Date" and "Value"). These rows will need to be excluded during the data cleaning stage before analysis.

- The dataset contains 2,843 rows and 2 columns.
- The first row describes the exchange-rate series.
- The second row contains the actual headers ("Date" and "Value").
- The exchange-rate observations appear to be daily.
- The data is from 02 January 2015 to 18 May 2026.
- The variable of the data is Rand per US Dollar ( USD/ZAR exchange rate).
- The exchange rates are ordered from the most recent to the oldest date.
- The dataset will require restructuring before it can be merged with the CPI datasets.
- The dataset is from SARB.

# Inspect EXCEL - CPI (COICOP 2018 - 8digit) (202604)

In [6]:
cpi_coicop.head(20)

,Division,DivisionDescription,Group,GroupDescription,Class,ClassDescription,Subclass,SubclassDescription,Old code,Eight digit code,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
0,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1111,Cereals,1111001.0,1111201,...,96.9,95.2,94.6,94.1,91.2,91.6,88.3,87.7,87.1,86.0
1,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1111,Cereals,NaN,1111202,...,93.1,93.2,93.8,94.3,93.8,92.6,91.9,91.0,92.1,91.9
2,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1112,Flour of cereals,1116001.0,1112101,...,101.4,101.8,101.9,101.5,101.4,100.9,101.4,100.8,101.3,101.0
3,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1112,Flour of cereals,1116002.0,1112102,...,102.1,100.8,102.7,100.9,103.1,100.0,99.7,99.6,99.5,99.6
4,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1112,Flour of cereals,NaN,1112301,...,99.7,101.2,102.0,102.6,103.4,103.2,103.0,103.4,103.6,104.3
5,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1112,Flour of cereals,1116003.0,1112601,...,111.1,110.5,111.3,110.1,109.2,109.5,107.5,108.1,106.3,105.6
6,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1112,Flour of cereals,1116012.0,1112602,...,112.2,113.0,113.1,113.0,112.8,112.4,111.6,111.6,111.0,110.9
7,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1113,Bread and bakery products,1112001.0,1113101,...,100.1,99.5,100.5,101.7,101.2,101.7,101.5,101.2,101.7,102.3
8,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1113,Bread and bakery products,1112002.0,1113102,...,99.4,100.0,100.6,100.9,101.1,101.0,100.8,100.4,100.4,100.5
9,1,Food and non-alcoholic beverages,11,Food,111,Cereal products,1113,Bread and bakery products,1112005.0,1113103,...,101.1,102.7,102.9,103.2,103.6,104.0,104.2,103.8,105.1,105.5


In [7]:
cpi_coicop.info()
cpi_coicop.describe()

<class 'pandas.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Columns: 233 entries, Division to M202604
dtypes: float64(221), int64(6), str(6)
memory usage: 711.9 KB


,Division,Group,Class,Subclass,Old code,Eight digit code,Weight,M200801,M200802,M200803,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
count,391.000000,391.000000,391.000000,391.000000,3.290000e+02,3.910000e+02,391.000000,218.000000,218.000000,218.000000,...,391.000000,391.000000,391.000000,391.000000,391.000000,391.000000,391.000000,391.000000,391.000000,391.000000
mean,4.516624,47.109974,473.923274,4741.552430,4.557498e+06,4.741748e+06,0.255754,56.694037,56.829358,57.318807,...,102.081841,102.298721,102.274425,102.258056,102.188491,102.034271,102.264706,102.489514,103.042967,103.596419
std,3.617494,36.811727,367.350980,3672.822072,3.590065e+06,3.672792e+06,0.741132,77.948724,77.375461,76.424219,...,6.543065,6.940053,6.909081,7.096137,6.238363,6.037949,6.575739,6.791512,7.146388,7.269729
min,1.000000,11.000000,111.000000,1111.000000,1.111001e+06,1.111201e+06,0.002322,13.500000,13.500000,13.500000,...,71.700000,62.300000,61.000000,59.300000,68.100000,76.400000,66.000000,58.100000,69.700000,71.500000
25%,1.000000,11.000000,118.000000,1183.000000,1.175002e+06,1.183501e+06,0.032057,32.100000,32.400000,32.775000,...,100.000000,100.150000,100.100000,100.100000,100.100000,100.000000,99.850000,100.050000,100.300000,100.500000
50%,4.000000,41.000000,411.000000,4110.000000,3.212002e+06,4.110101e+06,0.078589,38.450000,38.700000,39.650000,...,101.600000,101.900000,102.100000,101.900000,102.100000,102.000000,102.100000,102.200000,102.500000,103.500000
75%,7.000000,72.000000,721.000000,7212.000000,7.211001e+06,7.212104e+06,0.178207,52.275000,52.425000,52.625000,...,103.900000,104.200000,104.650000,104.800000,104.800000,104.850000,105.250000,105.700000,106.450000,107.250000
max,13.000000,139.000000,1390.000000,13909.000000,1.271100e+07,1.390991e+07,10.298931,1004.400000,995.200000,984.200000,...,172.100000,164.000000,164.400000,169.900000,142.400000,130.700000,132.100000,128.600000,153.400000,132.700000


In [8]:
cpi_coicop.tail()

,Division,DivisionDescription,Group,GroupDescription,Class,ClassDescription,Subclass,SubclassDescription,Old code,Eight digit code,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
386,13,Personal care and miscellaneous services,133,Social protection,1330,Social protection,13301,Child care services,NaN,13301201,...,106.8,106.8,106.8,106.8,106.8,106.8,106.8,106.8,113.5,113.5
387,13,Personal care and miscellaneous services,139,Other services,1390,Other services,13909,Other services n.e.c.,NaN,13909101,...,103.2,106.4,106.4,106.4,106.4,106.4,112.6,112.6,112.6,113.5
388,13,Personal care and miscellaneous services,139,Other services,1390,Other services,13909,Other services n.e.c.,NaN,13909401,...,100.0,100.0,100.0,100.2,100.2,100.2,100.8,100.8,100.8,105.1
389,13,Personal care and miscellaneous services,139,Other services,1390,Other services,13909,Other services n.e.c.,12711001.0,13909501,...,103.6,103.6,103.6,104.4,104.4,104.4,104.5,104.5,104.5,105.5
390,13,Personal care and miscellaneous services,139,Other services,1390,Other services,13909,Other services n.e.c.,12711002.0,13909911,...,100.7,100.7,100.7,101.2,101.2,101.2,101.7,101.7,101.7,101.7


In [9]:
cpi_coicop['GroupDescription'].unique()

<StringArray>
[                                                'Food',
                              'Non-alcoholic beverages',
                                  'Alcoholic beverages',
                                              'Tobacco',
                                             'Clothing',
                                             'Footwear',
                           'Actual rentals for housing',
                              'Owners' equivalent rent',
                               'Maintenance and repair',
              'Water supply and miscellaneous services',
                      'Electricty, gas and other fuels',
             'Furniture, furnishings and loose carpets',
                                   'Household textiles',
                                 'Household appliances',
          'Glassware, tableware and household utensils',
             'Tools and equipment for house and garden',
 'Goods and services for routine household maintenance',
                 

In [10]:
cpi_coicop['DivisionDescription'].unique()

<StringArray>
[                        'Food and non-alcoholic beverages',
                          'Alcoholic beverages and tobacco',
                                    'Clothing and footwear',
                                    'Housing and utilities',
 'Furnishings, household equipment and routine maintenance',
                                                   'Health',
                                                'Transport',
                            'Information and communication',
                            'Recreation, sport and culture',
                                       'Education services',
                   'Restaurants and accommodation services',
                         'Insurance and financial services',
                 'Personal care and miscellaneous services']
Length: 13, dtype: str

In [11]:
cpi_coicop['SubclassDescription'].unique()

<StringArray>
[                                                               'Cereals',
                                                       'Flour of cereals',
                                              'Bread and bakery products',
                                                      'Breakfast cereals',
                 'Macaroni, noodles, couscous and similar pasta products',
                                         'Meat, fresh, chilled or frozen',
                               'Meat , dried, salted, in brine or smoked',
                    'Offal, blood and other parts of slaughtered animals',
 'Meat, offal, blood and other parts of slaughtered animals preparations',
                                                                   'Fish',
 ...
                                            'Life and accident insurance',
                                        'Insurance connected with health',
                                  'Insurance connected with the dwelling',
      

### Observations of the CPI (COICOP 2018 - 8digit) (202604) dataset

- The dataset contains 391 observations and 233 variabes.
- The variables consist of 221 floating-point, 6 integer, and 6 string data types.
- The first seven columns contain classification information, including Division, Group, Class, Subclass, product codes, and CPI weights.
- The dataset is structured hierarchically using the COICOP 2018 classification system, consisting of 13 divisions, 44 groups, and numerous detailed subclasses.
- The DivisionDescription variable identifies broad expenditure categories, while GroupDescription, ClassDescription, and SubclassDescription provide increasingly detailed classifications.
- Monthly CPI index values are recorded from January 2008 (M200801) to April 2026 (M202604).
- The dataset includes detailed food-related subclasses such as cereals, flour of cereals, bread and bakery products, meat, and fish, which align directly with the focus of this research.
- The dataset originates from Statistics South Africa (Stats SA) and provides sufficient historical coverage for the study period.
- The dataset contains all CPI divisions but the study is specifically about food. One of the first preprocessing steps will be to filtwr the dataset to that.

## Inspect CPI_Average prices_Provinces(202604)

In [12]:
prices_provinces.head(20)

,H01,H02,Old code,Old description,H03,H04,H05,H06,H07,H08,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
0,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,31.87,32.87,32.37,29.74,29.12,27.87,26.49,28.74,28.12,27.32
1,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,28.88,28.88,28.74,28.62,27.99,26.09,26.49,25.99,26.88,26.77
2,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,28.82,29.15,28.65,28.65,27.32,28.38,28.38,27.98,27.98,27.98
4,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,32.21,32.21,32.21,32.29,31.89,32.09,31.29,31.59,31.59,28.29
5,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,29.59,29.59,29.59,29.59,29.59,29.49,29.17,27.54,26.99,26.99
7,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,27.85,26.56,26.56,27.28,26.13,25.42,25.16,26.82,24.49,24.49
8,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10924,2 Kilogram,...,47.32,46.17,46.63,45.61,41.84,44.32,42.07,43.53,43.32,39.99
9,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10924,2 Kilogram,...,44.99,42.80,42.93,40.11,39.47,40.26,37.99,38.27,37.49,38.15


In [13]:
prices_provinces.info()
prices_provinces.describe()

<class 'pandas.DataFrame'>
RangeIndex: 2455 entries, 0 to 2454
Columns: 124 entries, H01 to M202604
dtypes: float64(113), int64(3), str(8)
memory usage: 2.3 MB


,Old code,H03,H07,Province_code,M201701,M201702,M201703,M201704,M201705,M201706,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
count,1.274000e+03,2.455000e+03,2455.000000,2455.000000,1135.000000,1126.000000,1130.000000,1132.000000,1133.000000,1132.000000,...,2326.000000,2324.000000,2319.000000,2292.000000,2293.000000,2286.000000,2283.000000,2271.000000,2274.000000,2269.000000
mean,2.248150e+06,2.717075e+06,10885.348676,5.052953,41.953013,42.003757,42.202982,42.183348,42.814042,42.938189,...,145.768779,145.897978,145.527451,144.538329,144.949398,142.971942,144.625112,144.143950,144.702427,145.192226
std,2.749776e+06,3.018871e+06,137.865235,2.520585,42.041371,41.964102,42.183736,42.190437,43.374024,43.196479,...,526.973025,532.756055,531.937777,529.009655,523.473004,496.583401,517.738675,517.965047,514.628649,511.198805
min,1.111001e+06,1.111201e+06,10508.000000,1.000000,3.270000,3.330000,3.440000,3.330000,3.600000,3.600000,...,2.750000,2.750000,2.750000,2.750000,2.770000,2.770000,2.760000,2.760000,2.760000,2.770000
25%,1.124002e+06,1.133901e+06,10820.000000,3.000000,15.185000,15.212500,15.090000,15.362500,15.300000,15.325000,...,28.560000,27.990000,27.555000,27.897500,27.790000,27.802500,27.495000,27.545000,27.882500,28.130000
50%,1.173001e+06,1.181101e+06,10902.000000,5.000000,26.490000,26.410000,26.270000,25.980000,26.230000,26.615000,...,48.870000,48.835000,49.370000,49.545000,49.290000,49.130000,49.290000,48.730000,48.685000,49.280000
75%,1.211752e+06,3.124101e+06,10963.000000,7.000000,47.420000,46.112500,47.327500,46.457500,46.490000,47.397500,...,100.275000,100.795000,102.620000,103.795000,104.090000,104.982500,106.980000,106.085000,105.530000,106.660000
max,1.213101e+07,1.312030e+07,11445.000000,9.000000,290.560000,290.560000,298.980000,288.730000,296.640000,327.490000,...,9532.330000,9249.120000,9832.440000,10927.570000,11332.330000,8999.240000,11499.240000,10427.710000,9587.990000,10279.100000


In [14]:
prices_provinces.tail()

,H01,H02,Old code,Old description,H03,H04,H05,H06,H07,H08,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
2450,P0141,Consumer Price Index,12131008.0,Skin Lotion,13120302,Skin lotion,Price (Rand),Monthly,11034,400 Millilitre,...,58.89,57.86,56.18,53.24,54.41,53.30,52.76,52.89,53.60,53.22
2451,P0141,Consumer Price Index,12131008.0,Skin Lotion,13120302,Skin lotion,Price (Rand),Monthly,11034,400 Millilitre,...,49.16,47.66,47.66,48.49,47.99,47.99,47.57,47.66,47.82,48.16
2452,P0141,Consumer Price Index,12131008.0,Skin Lotion,13120302,Skin lotion,Price (Rand),Monthly,11034,400 Millilitre,...,55.12,55.31,55.78,55.53,54.85,55.33,55.72,55.80,55.51,55.56
2453,P0141,Consumer Price Index,12131008.0,Skin Lotion,13120302,Skin lotion,Price (Rand),Monthly,11034,400 Millilitre,...,46.40,45.12,46.71,47.16,46.62,46.66,46.72,46.89,45.27,48.40
2454,P0141,Consumer Price Index,12131008.0,Skin Lotion,13120302,Skin lotion,Price (Rand),Monthly,11034,400 Millilitre,...,32.90,32.44,32.90,33.54,33.54,33.54,33.54,33.72,33.81,34.81


#### Observations of the CPI_Average prices_Provinces(202604) dataset

- The dataset contains 2,455 observations and 124 variables.
- Monthly average retail prices are reported from January 2017 through April 2026.
- Unlike the COICOP dataset, this dataset records actual prices in South African Rand rather than CPI index values.
- Each observation represents a specific product measured using a defined unit (for example, White Rice – 1 Kilogram).
- Several observations contain missing monthly values, indicating incomplete price records for some products or provinces.
- The dataset appears to contain repeated products, suggesting that prices are recorded separately across provinces.
- The dataset originates from Statistics South Africa.

## Inspect CPI_Average Prices_All urban(202604)

In [15]:
prices_urban.head(20)

,H01,H02,Old code,Old description,H03,H04,H05,H06,H07,H08,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
0,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10917,1 Kilogram,...,29.59,29.59,29.45,29.29,28.81,28.39,28.26,28.12,27.82,27.06
1,P0141,Consumer Price Index,1111001.0,Rice,1111201,White Rice,Price (Rand),monthly,10924,2 Kilogram,...,43.37,42.66,42.45,41.92,40.68,41.16,39.46,39.00,38.78,38.13
2,P0141,Consumer Price Index,NaN,NaN,1111202,Basmati Rice,Price (Rand),monthly,10917,1 Kilogram,...,74.06,74.03,74.68,75.75,74.49,73.22,73.18,71.91,72.62,73.29
3,P0141,Consumer Price Index,NaN,NaN,1111202,Basmati Rice,Price (Rand),monthly,10924,2 Kilogram,...,132.23,133.48,134.69,133.25,131.83,133.72,134.18,135.06,133.29,133.74
4,P0141,Consumer Price Index,NaN,NaN,1112101,Cake flour,Price (Rand),monthly,10917,1 Kilogram,...,21.40,21.26,21.20,21.36,21.41,21.21,21.34,21.36,21.34,21.50
5,P0141,Consumer Price Index,1116001.0,Cake flour,1112101,Cake flour,Price (Rand),monthly,10926,2.5 Kilogram,...,42.57,42.91,42.95,42.59,42.68,42.45,42.71,42.26,42.45,42.62
6,P0141,Consumer Price Index,NaN,NaN,1112102,Bread flour,Price (Rand),monthly,10917,1 Kilogram,...,22.28,21.40,21.80,22.14,21.77,21.75,22.32,22.21,22.44,21.48
7,P0141,Consumer Price Index,1116002.0,Bread flour,1112102,Bread flour,Price (Rand),monthly,10926,2.5 Kilogram,...,46.67,47.24,48.03,48.43,47.83,47.19,46.65,46.68,46.84,46.91
8,P0141,Consumer Price Index,NaN,NaN,1112301,Sorghum Meal,Price (Rand),monthly,10917,1 Kilogram,...,30.29,30.34,29.96,30.48,31.16,30.83,30.56,30.52,30.36,31.28
9,P0141,Consumer Price Index,NaN,NaN,1112301,Sorghum Meal,Price (Rand),monthly,10924,2 Kilogram,...,54.67,55.78,56.37,57.19,57.50,57.14,56.94,57.20,57.26,57.35


In [16]:
prices_urban.info()
prices_urban.describe()

<class 'pandas.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Columns: 123 entries, H01 to M202604
dtypes: float64(114), int64(2), str(7)
memory usage: 375.9 KB


,Old code,H03,H07,M201701,M201702,M201703,M201704,M201705,M201706,M201707,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
count,1.670000e+02,3.910000e+02,391.000000,155.000000,155.000000,155.000000,155.000000,155.000000,155.000000,155.000000,...,390.000000,389.000000,390.000000,390.000000,390.000000,390.000000,390.000000,389.000000,389.000000,389.000000
mean,2.300592e+06,2.760485e+06,10889.721228,42.096774,42.352452,42.647355,42.624194,43.156065,43.101806,43.453742,...,260.155333,258.890925,261.072487,257.729385,255.834718,252.835026,255.587538,255.625064,253.988560,252.828355
std,2.798993e+06,2.946419e+06,149.896807,41.246878,41.587712,41.891372,42.113408,43.210988,42.644559,43.364997,...,1031.389756,1021.777365,1047.768251,1026.614346,1006.143627,981.317270,994.477780,987.112805,967.280775,957.723168
min,1.111001e+06,1.111201e+06,10508.000000,3.740000,3.760000,3.810000,3.790000,3.920000,3.980000,4.000000,...,3.190000,3.390000,3.390000,3.400000,3.570000,3.620000,3.330000,3.310000,3.340000,3.300000
25%,1.125004e+06,1.142002e+06,10813.500000,15.415000,15.505000,14.955000,15.110000,15.295000,15.015000,14.690000,...,27.990000,27.840000,27.000000,27.647500,27.740000,27.757500,27.827500,28.120000,27.860000,28.240000
50%,1.172001e+06,1.184001e+06,10897.000000,26.720000,26.480000,26.420000,26.040000,26.370000,26.480000,26.770000,...,51.390000,50.750000,51.630000,51.180000,51.100000,50.990000,51.370000,51.230000,51.580000,52.650000
75%,1.211502e+06,3.124105e+06,10963.000000,48.425000,49.090000,49.025000,48.975000,49.965000,49.845000,50.160000,...,115.982500,116.370000,117.232500,117.590000,119.732500,121.130000,123.750000,120.990000,120.550000,120.930000
max,1.213101e+07,1.312030e+07,11445.000000,249.080000,249.690000,253.380000,253.150000,270.080000,252.170000,263.070000,...,11416.380000,11295.480000,12214.030000,12258.440000,11899.160000,11216.560000,11478.370000,11206.100000,10409.230000,10508.870000


In [17]:
prices_urban.tail()

,H01,H02,Old code,Old description,H03,H04,H05,H06,H07,H08,...,M202507,M202508,M202509,M202510,M202511,M202512,M202601,M202602,M202603,M202604
386,P0141,Consumer Price Index,12131012.0,Tooth Paste,13120205,Toothpaste,Price (Rand),Monthly,11008,100 Millilitre,...,30.06,30.28,30.56,30.31,30.19,30.34,30.34,30.28,30.14,30.04
387,P0141,Consumer Price Index,12131004.0,Sanitory Pads,13120206,Sanitary pads,Price (Rand),Monthly,10740,Per Pack,...,28.84,28.83,29.30,29.63,29.94,29.65,29.64,29.66,29.35,29.04
388,P0141,Consumer Price Index,NaN,NaN,13120301,Deodorant,Price (Rand),monthly,11000,50 Millilitre,...,31.83,31.76,31.89,32.02,32.03,32.04,31.84,32.09,31.92,32.01
389,P0141,Consumer Price Index,12131013.0,Aerosol Deodorant,13120301,Deodorant,Price (Rand),Monthly,11011,150 Millilitre,...,46.91,46.87,48.17,47.86,48.49,48.65,48.76,48.11,48.30,48.68
390,P0141,Consumer Price Index,12131008.0,Skin Lotion,13120302,Skin lotion,Price (Rand),Monthly,11034,400 Millilitre,...,50.26,49.71,50.17,50.37,50.37,50.55,50.65,51.12,51.07,51.73


#### Observations of CPI_Average Prices_All urban(202604) dataset
- The dataset contains 391 observations and 123 variables.
- It records monthly average retail prices for a wide range of consumer products from January 2017 (M201701) to April 2026 (M202604).
- Similar to the CPI_Average prices_Provinces dataset, this dataset contains actual average retail prices in South African Rand.
- Each observation represents a specific product together with its unit of measurement (e.g., White Rice – 1 Kilogram, Cake Flour – 2.5 Kilograms).
- Some variables contain missing values, particularly the Old code field and certain monthly observations, indicating that not all products have complete historical records.
- The dataset represents national average (All Urban) prices, whereas the provincial dataset provides prices at the provincial level.
- The dataset originates from Statistics South Africa (Stats SA).


| Dataset                         | Frequency | Measures                       | Research Role                  |
| ------------------------------- | --------- | ------------------------------ | ------------------------------ |
| **HistoricalRateDetail (SARB)** | Daily     | USD/ZAR exchange rate          | Independent variable           |
| **COICOP 2018 (8-digit)**       | Monthly   | CPI index by food category     | **Primary dependent variable** |
| **Average Prices (All Urban)**  | Monthly   | National average retail prices | Supporting dataset             |
| **Average Prices (Provinces)**  | Monthly   | Provincial retail prices       | Extension dataset   |


Four datasets were examined for their relevance to the research objectives. The SARB HistoricalRateDetail dataset provides daily USD/ZAR exchange rates, while the COICOP 2018 (8-digit) dataset provides monthly CPI indices across detailed food categories. These two datasets are selected as the primary datasets because they directly support the modelling of exchange rate pass-through into food prices. The Average Prices (All Urban) and Average Prices (Provinces) datasets provide complementary information on retail prices and may be used for descriptive analysis.